ADApp/ADSrc/NDArrayPool.cpp

```diff
@@ -121,156 +98,97 @@ NDArray* NDArrayPool::getNextFreeArray(NDArray *pArray)
   */
 NDArray* NDArrayPool::alloc(int ndims, size_t *dims, NDDataType_t dataType, size_t dataSize, void *pData)
 {
-  NDArray *pArray=NULL, *freeArray=NULL, *nextArray=NULL, *closestArray=NULL, *threshArray=NULL;
-
+  NDArray *pArray=NULL;
   NDArrayInfo_t arrayInfo;
-  int i;
   const char* functionName = "NDArrayPool::alloc:";
 
   epicsMutexLock(listLock_);
 
-  /* Find a free image */
-  freeArray = getFirstFreeArray();
+  // Compute the required NDArray size
+  NDArray::computeArrayInfo(ndims, dims, dataType, &arrayInfo);
+  if (dataSize == 0) {
+    dataSize = arrayInfo.totalBytes;
+  } 
+  else if (dataSize < arrayInfo.totalBytes) {
+      printf("%s: ERROR: required size=%d passed size=%d is too small, using required size\n",
```

Comment:
Should this be an asynPrintf call or removed?

Reply:
There is no asynUser available in NDArrayPool.  There is potentially NDArrayPool::pDriver_, which points to an asynNDArrayDriver.  However, that is a forward referenced class in NDArrayPool.  Furthermore the asynNDArrayDriver::pasynUserSelf member is protected, so NDArrayPool does not have access to it.

Reply:
In this case would it then be better to use the EPICS logging framework instead: errMessage, errPrintf, or epicsPrintf?

Reply:
I have made the changes in NDArray.h, asynNDArrayDriver.h, and NDArrayPool.cpp to use asynPrint rather than printf.   The unit test was already sufficient to test that this works.